Services/source/cli/fs_utils.c

```diff
@@ -0,0 +1,330 @@
+        break;
+    case 27:
+        errorMsg = "RED_EFBIG";
+        break;
+    case 28:
+        errorMsg = "RED_ENOSPC";
+        break;
+    case 30:
+        errorMsg = "RED_EROFS";
+        break;
+    case 31:
+        errorMsg = "RED_EMLINK";
+        break;
+    case 34:
+        errorMsg = "RED_ERANGE";
+        break;
+    case 36:
+        errorMsg = "RED_ENAMETOOLONG";
+        break;
+    case 38:
+        errorMsg = "RED_ENOSYS";
+        break;
+    case 39:
+        errorMsg = "RED_ENOTEMPTY";
+        break;
+    case 61:
+        errorMsg = "RED_ENODATA";
+        break;
+    case 87:
+        errorMsg = "RED_EUSERS";
+        break;
+    default:
+        errorMsg = "Unknown Error";
+    }
+    snprintf(pcWriteBuffer, xWriteBufferLen, "%s\n", errorMsg);
+    red_errno = 0;
+}
+
+static BaseType_t prvPWDCommand(char *pcWriteBuffer, size_t xWriteBufferLen, const char *pcCommandString) {
+    char pszBuffer[REDCONF_NAME_MAX];
+    char *cwd = red_getcwd(pszBuffer, REDCONF_NAME_MAX);
+
+    if (cwd == NULL) {
+        createErrorOutput(pcWriteBuffer, xWriteBufferLen);
+    } else {
+        snprintf(pcWriteBuffer, xWriteBufferLen, "%s\n", cwd);
+    }
+    return pdFALSE;
+}
+
+static BaseType_t prvCDCommand(char *pcWriteBuffer, size_t xWriteBufferLen, const char *pcCommandString) {
+    // We can guarantee there will be one parameter because FreeRTOS+CLI won't call this function unless it has
+    // exactly one parameter
+    BaseType_t parameterLen;
+    const char *parameter = FreeRTOS_CLIGetParameter(
+        /* The command string itself. */
+        pcCommandString,
+        /* Return the first parameter. */
+        1,
+        /* Store the parameter string length. */
+        &parameterLen);
+    int32_t error = red_chdir(parameter);
+    if (error < 0) {
+        createErrorOutput(pcWriteBuffer, xWriteBufferLen);
+    }
+    return pdFALSE;
+}
+
+static BaseType_t prvLSCommand(char *pcWriteBuffer, size_t xWriteBufferLen, const char *pcCommandString) {
+    static bool firstRun = true;
+    static REDDIR *dir;
+    if (firstRun) {
+        red_errno = 0;
+        char pszBuffer[REDCONF_NAME_MAX];
+        char *cwd = red_getcwd(pszBuffer, REDCONF_NAME_MAX);
+        if (cwd == NULL) {
+            createErrorOutput(pcWriteBuffer, xWriteBufferLen);
+            return pdFALSE;
+        }
+        dir = red_opendir(cwd);
+        if (dir == NULL) {
+            createErrorOutput(pcWriteBuffer, xWriteBufferLen);
+            return pdFALSE;
+        }
+        firstRun = false;
+    }
+    red_errno = 0;
+    REDDIRENT *current = red_readdir(dir);
+    if (current == NULL) {
+        if (red_errno == 0) {
+            snprintf(pcWriteBuffer, xWriteBufferLen, "\n");
+        } else {
+            createErrorOutput(pcWriteBuffer, xWriteBufferLen);
+        }
+        red_closedir(dir);
+        firstRun = true;
+        return pdFALSE;
+    } else {
+        snprintf(pcWriteBuffer, xWriteBufferLen, "%s\n", current->d_name);
```

Comment:
Is this "ls" or "pwd"? I would have thought "ls" would print all the files in the directory vs the directory name?

Reply:
REDDIRENT stores both files and directories. I'm grabbing the name from that struct